### PR TITLE
phan: Add a plugin for handling "html template" files

### PR DESCRIPTION
**Pull request by anomiex**
## Proposed changes:

An "html template" file is a PHP file that produces a fragment of HTML, loaded by a method or function call somewhere else in the codebase.

Phan will normally evaluate the template as being in the global scope, meaning it doesn't know about `$this` or other variables and will complain about attempts to access them.

This plugin adds support for `@html-template` and `@html-template-var` annotations to declare that a file is a template, what its context is, and what variables are intended to be available.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

None

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

* CI happy?
* Does all this make sense?

### CODE REVIEW

File: projects/packages/phan-plugins/src/HtmlTemplatePlugin.php [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+use Phan\PluginV3;
+use Phan\PluginV3\BeforeAnalyzeFileCapability;
+use PhpToken;
+
+/**
+ * Phan plugin for handling "html template" files.
+ */
+class HtmlTemplatePlugin extends PluginV3 implements BeforeAnalyzeFileCapability {
+
+	private const IDENTIFIER  = '[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*';
```
Comment (tbradsha) [inline projects/packages/phan-plugins/src/HtmlTemplatePlugin.php:30]: Tiny nitpick: I suspect we don't want to include `\x7f` (`DEL`). PHP documentation seems to confirm that:
https://www.php.net/manual/en/language.variables.basics.php

```suggestion
	private const IDENTIFIER  = '[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*';
```

File: projects/plugins/jetpack/modules/tiled-gallery/tiled-gallery/tiled-gallery-layout.php [hunk 2]
```diff
@@ -130,10 +132,12 @@ private function template( $name, $context = null ) {
 	/**
 	 * Handle tiled gallery partial path.
 	 *
+	 * @html-template-var array $context
+	 *
 	 * @param string $name - the name.
 	 * @param array  $context Context array passed to the partial.
 	 */
-	private function partial( $name, $context = null ) {
```
Comment (tbradsha) [inline projects/plugins/jetpack/modules/tiled-gallery/tiled-gallery/tiled-gallery-layout.php:136]: I was going to see if this and the next default param change affected the filter docs, but it doesn't look like the filters are documented. :^)

File: projects/packages/phan-plugins/src/HtmlTemplatePlugin.php (hunk 1 above)
Comment (anomiex) [reply projects/packages/phan-plugins/src/HtmlTemplatePlugin.php:30]: I copied the regex from Phan's own code. Turns out it was allowed [until PHP 7.1](https://www.php.net/manual/en/migration71.incompatible.php#migration71.incompatible.delete-control-character-in-identifiers).

Comment (tbradsha) [reply projects/packages/phan-plugins/src/HtmlTemplatePlugin.php:30]: So you mean I spent all that time reviewing the regex for naught? 😆

Comment (anomiex) [reply projects/packages/phan-plugins/src/HtmlTemplatePlugin.php:30]: You might have caught a bug in Phan by reviewing it. 😀

**Review by tbradsha — COMMENTED** [review]
I did a high-level pass and things look pretty good, but it might be good to have a more experienced set of eyes on this one.

Two thoughts while reviewing:
* Could we parse the already-existing `@param` rather than needing to define our own `@html-template-var` in instances where we need it for the template?
* Perhaps more difficult to address, but when defining a `@html-template-var`, it assumes the variable becomes available at any point in the scope. This could mean we're missing issues where the actual var is defined after the template inclusion.

On that last point, the above isn't very clear. Take this variation of one of your tests, it would act like there's no issue:

```
/**
 * @html-template-var 'ns/f2' $var
 */
function loadTemplate2( $name ) {
	require __DIR__ . "/templates/$name.php";
	$var = (object) array();
}
```

**Review by tbradsha — APPROVED** [review]

**Comment by anomiex** [comment]
> * Could we parse the already-existing `@param` rather than needing to define our own `@html-template-var` in instances where we need it for the template?

We could, if we want to assume that any `@param` is intended to be passed on to the template. Even if it accidentally is, I'm not sure it would be intended.

> * Perhaps more difficult to address, but when defining a `@html-template-var`, it assumes the variable becomes available at any point in the scope. This could mean we're missing issues where the actual var is defined after the template inclusion.

If it's defined after the template inclusion, it shouldn't be declared with `@html-template-var` because it's not a variable for use by the template.

**Comment by tbradsha** [comment]
> We could, if we want to assume that any @param is intended to be passed on to the template. Even if it accidentally is, I'm not sure it would be intended.

Intentional or not, it would more accurately reflect the reality of what's in the scope (same as if it were all inline as part of the same file).

>  If it's defined after the template inclusion, it shouldn't be declared with `@html-template-var` because it's not a variable for use by the template.

I'm mainly thinking of a scenario with a more complex function, where the `@html-template-var` is already defined and then the parent function logic gets rearranged. Below is the edge case scenario I'm describing:

Original code:
```
/**
 * @html-template-var 'ns/f2' $var
 */
function loadTemplate2( $name ) {
	$var = (object) array();
	// lots of code here
	require __DIR__ . "/templates/$name.php";
}
```

Future code:
```
/**
 * @html-template-var 'ns/f2' $var
 */
function loadTemplate2( $name ) {
	// lots of code here
	require __DIR__ . "/templates/$name.php";
	// lots of code here
	$var = (object) array();
}
```

Basically once it's defined, it's permanently regarded by Phan as being a valid function within the template, whether or not that continues to hold true. If instead we could do an inline comment at the point the var is initialized, that would mean at any point earlier the var would not be assumed valid by Phan. An inline comment is more likely to be moved around with its associated variable initialization.

This is what I'm theoretically suggesting, though I'm not even sure if that's possible or if Phan would even work that way:
```
function loadTemplate2( $name ) {
	// @html-template-var 'ns/f2' $var
	$var = (object) array();
	// lots of code here
	require __DIR__ . "/templates/$name.php";
}
```

**Comment by anomiex** [comment]
> Intentional or not, it would more accurately reflect the reality of what's in the scope (same as if it were all inline as part of the same file).

Unless the code `unset()` the parameter, or reset it to some other value, or something. But beyond that, I'm looking at what someone wants to document is intended to be available, not what accidentally is also available because they left out an `unset()`.

> Basically once it's defined, it's permanently regarded by Phan as being a valid function within the template, whether or not that continues to hold true.

What you're saying here strikes me as the same thing as a case like this:
```
/**
 * @param bool $something
 */
function do_a_thing( $something ) {
    if ( is_wp_error( $something ) ) {
        return $something;
    }
   // Do some stuff.
}
```
The fix is to fix the definition to match reality. Same here, if `$var` is no longer available to the template then you should remove the `@html-template-var type $var` to match reality.

> This is what I'm theoretically suggesting, though I'm not even sure if that's possible or if Phan would even work that way

I suppose it might be possible, but it would be much more complicated complicated. First, you'd have to do like `'@html-template-var ...';` instead of `// @html-template-var ...` because the AST only keeps doc comments on classes, functions, and such, inline comments are thrown away. Or else you'd need to somehow go from the AST node back to the source file, which would be even more of a pain. And then you'd have to process the function during the parse to collect those, and also manage to determine where the `require` inside the function is. OTOH, once you did all that work you could probably just do `'@html-template-var $var';` and let Phan determine the type of `$var` at the point of the `require`.

And then consider that you'd lose all that if you have to make stubs for the function instead of parsing the actual code. 🤷
